blog: announce Codemod Registry for Backstage upgrades (#35015)
backstage/backstage: microsite/blog/2026-07-28-automating-backstage-upgrades-with-codemods.mdx

---
title: Automating Backstage upgrades with codemods
author: Paul Schultz, Red Hat
authorURL: https://github.com/schultzp2020
authorImageURL: https://avatars.githubusercontent.com/u/52387464?v=4
---

> **TL;DR:** Backstage now publishes official codemod recipes on the [Codemod Registry](https://go.codemod.com/backstage). After you bump packages and apply create-app diffs, run the *versioned* recipe for your target release. Use *misc* recipes, like Material-UI to Backstage UI, when you are ready for a bigger migration.

Reviewer: Can we include a link to both the Codemod Registry and the repo? I've had Adopters mention that they sometimes have issues with 3rd part tools that get blocked at their Org and I think only having a link to Codemod Registry gives the impression that you have to interact with them that way when that's not the case.

> **TL;DR:** Backstage now publishes official codemod recipes on the [Codemod Registry](https://go.codemod.com/registry). After you bump packages and apply create-app diffs, run the **versioned** recipe for your target release — and reach for **misc** recipes, like Material-UI to Backstage UI, when you are ready for a bigger migration.

Reviewer 2: if you want to send folks directly to Backstage codemods you can use this:

https://go.codemod.com/backstage

cc @schultzp2020

> **TL;DR:** Backstage now publishes official codemod recipes on the [Codemod Registry](https://go.codemod.com/registry). After you bump packages and apply create-app diffs, run the **versioned** recipe for your target release — and reach for **misc** recipes, like Material-UI to Backstage UI, when you are ready for a bigger migration.

Reviewer: ```suggestion
**TL;DR:** Backstage now publishes official codemod recipes on the [Codemod Registry](https://go.codemod.com/registry). After you bump packages and apply `create-app` diffs, run the **versioned** recipe for your target release — and reach for **misc** recipes, like Material UI (MUI) to Backstage UI (BUI), when you are ready for a bigger migration.
```

Reviewer: I remember that Material UI was very specific about this. If you look into the PRs there was a big one they created to remove the `-` from all mentions in the Backstage code base and docs.

{/* truncate */}

## Upgrades have two automation layers

If you have upgraded a Backstage app more than once, you already know the rhythm: run `yarn backstage-cli versions:bump`, open the [Backstage Upgrade Helper](https://backstage.github.io/upgrade-helper/), and apply the `create-app` template changes that do not land automatically in your `app` and `backend` packages.

> That first layer still matters. Package bumps and template diffs answer "what should my scaffolding look like for this release?" They leave your own plugins alone: deprecated props, config keys, imports. That cleanup has always been a second job.

Reviewer: ```suggestion
Those initial steps still matter. Package bumps and template diffs answer "what should my scaffolding look like for this release?" They leave your own plugins alone: deprecated props, config keys, imports. That cleanup has always been a second job.
```

> The second layer is new. The [backstage/codemods](https://github.com/backstage/codemods) repository publishes recipes (ordered bundles of source transforms) as `@backstage/*` packages on the [Codemod Registry](https://go.codemod.com/backstage). You run them with the [Codemod CLI](https://docs.codemod.com/cli), typically via `yarn dlx codemod run …`, and they rewrite that plugin code for you.

Reviewer: This phrase - "The second layer is new." - seems out of place. You don't mention there is more than one layer until this point. Maybe we should foreshadow this in the prior paragraphs?

> The second layer is new. The [backstage/codemods](https://github.com/backstage/codemods) repository publishes recipes (ordered bundles of source transforms) as `@backstage/*` packages on the [Codemod Registry](https://go.codemod.com/backstage). You run them with the [Codemod CLI](https://docs.codemod.com/cli), typically via `yarn dlx codemod run …`, and they rewrite that plugin code for you.

Reviewer: In the repo docs we use `npx` - https://github.com/backstage/codemods - but here we use `dlx`, can we just use `npx` for consistency?

Author: We cannot use `npx` because some commands (like codemods) break when in a yarn repo. I can investigate to see if there are some inconsistencies and clean them up to use `dlx`.

Reviewer: Ah, then should the [`backstage/codemods`](https://github.com/backstage/codemods#usage) repo then be updated?

> The second layer is new. The [backstage/codemods](https://github.com/backstage/codemods) repository publishes recipes — ordered bundles of source transforms — as `@backstage/*` packages on the [Codemod Registry](https://go.codemod.com/registry). You run them with the [Codemod CLI](https://docs.codemod.com/cli), typically via `yarn dlx codemod@latest run …`, and they rewrite that plugin code for you.

Reviewer 2: not need to mention @latest after codemod (we have built in feature that prevents folks running stale codemods). that makes the command look nicer and easier to read. @schultzp2020

> The [backstage/codemods](https://github.com/backstage/codemods) repository publishes recipes (ordered bundles of source transforms) as `@backstage/*` packages on the [Codemod Registry](https://go.codemod.com/backstage). You run them with the [Codemod CLI](https://docs.codemod.com/cli), typically via `yarn dlx codemod run …`, and they rewrite that plugin code for you.

Reviewer: ```suggestion
What's new is the [backstage/codemods](https://github.com/backstage/codemods) repository which publishes recipes (ordered bundles of source transforms) as `@backstage/*` packages on the [Codemod Registry](https://go.codemod.com/backstage). You run them with the [Codemod CLI](https://docs.codemod.com/cli), typically via `yarn dlx codemod run …`, and they rewrite that plugin code for you!
```

> These recipes are not the old `@backstage/codemods` npm package.

Reviewer: I'm not sure this makes sense to mention at all. I really don't think people have been using it. If anything you might want to add this as a line in the new dedicated repo but for this blog post I think it's best not to mention it at all.

> These recipes are separate from the legacy `@backstage/codemods` npm package, which has been removed from the main monorepo. If you still depend on its old jscodeshift transforms, pin the last published version of that package on npm.

Reviewer: Do you think we should really mention this? The weekly downloads are somewhat high, but historically they have been pretty low. I wonder if AI tools are looking at it but end up finding it doesn't help, could be why the numbers are up - https://www.npmjs.com/package/@backstage/codemods

## Versioned recipes: part of the upgrade path

Recipes come in two groups, and the first is tied to releases. For a target release `1.XX.0`, look for `@backstage/v1-XX-0-migration-recipe`. For example, when moving toward 1.52.0:

```shell
yarn dlx codemod run @backstage/v1-52-0-migration-recipe \
  --target . \
  --dry-run

yarn dlx codemod run @backstage/v1-52-0-migration-recipe \
  --target .
```

Dry-run first, then apply. Afterwards, search your repo for `TODO(backstage-codemod)` markers, and check the recipe's README for out-of-scope items you will need to change by hand. If a release has no recipe yet, skip this step.

Make it a standard part of the routine: bump, Upgrade Helper, then codemods. The [Keeping Backstage Updated](https://backstage.io/docs/getting-started/keeping-backstage-updated) guide walks through the full flow.

## Misc recipes: opt-in migrations

> Not every migration belongs to a single monthly release. *Misc* recipes cover larger, cross-cutting work you schedule when your team is ready. The first misc recipe available is the Material-UI 4 to Backstage UI migration:

Reviewer: ```suggestion
Not every migration belongs to a single monthly release. *Misc* recipes cover larger, cross-cutting work you schedule when your team is ready. The first misc recipe available is the Material UI 4 to Backstage UI migration:
```
